### PR TITLE
SONARJAVA-1747 Adding support for Guava VisibleForTesting annotation

### PR DESCRIPTION
**Pull request by m-g-sonar**
Squash commits of @cardil from PR https://github.com/SonarSource/sonar-java/pull/848 into a single one and slightly rework rule descriptions + fix quality flaw.

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java [hunk 1]
```diff
@@ -47,29 +52,39 @@
...
+  private void checkComplianceOnModifiersAndSymbol(ModifiersTree modifiers, Symbol symbol) {
     ModifierKeywordTree modifier = ModifiersUtils.getModifier(modifiers, Modifier.PROTECTED);
-    if (modifier != null) {
-      reportIssue(modifier.keyword(), "Remove this \"protected\" modifier.");
+    if (modifier != null && !isVisibleForTesting(symbol)) {
+      reportIssue(modifier.keyword(), MESSAGE);
     }
   }
 
+  private boolean isVisibleForTesting(Symbol symbol) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java]: MAJOR Make "isVisibleForTesting" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java [hunk 2]
```diff
@@ -47,29 +52,39 @@
...
     ClassTree classTree = (ClassTree) tree;
     if (ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
-      for (Tree member : classTree.members()) {
-        checkMember(member);
-      }
+      classTree.members().forEach(this::checkMember);
     }
   }
 
   private void checkMember(Tree member) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java]: MAJOR Private method 'checkMember' is never used. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java [hunk 3]
```diff
@@ -43,18 +44,28 @@
...
         IdentifierTree simpleName = ((VariableTree) member).simpleName();
         reportIssue(simpleName, "Explicitly declare the visibility for \"" + simpleName.name() + "\".");
-      }
-    }
+      });
   }
 
-  private static boolean hasNoVisibilityModifier(VariableTree member) {
-    ModifiersTree modifiers = member.modifiers();
+  private static boolean isConsentWithCheck(Tree member) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java]: MAJOR Private method 'isConsentWithCheck' is never used. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java (hunk 3 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java]: FP. will be handled by resolution of [SONARJAVA-958](https://jira.sonarsource.com/browse/SONARJAVA-958)

File: java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java (hunk 2 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ProtectedMemberInFinalClassCheck.java]: FP. will be handled by resolution of [SONARJAVA-958](https://jira.sonarsource.com/browse/SONARJAVA-958)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by cardil** [comment]
Nice! :smiley:

**Pull request merged**